snuba/pipeline/stages/query_execution.py

```diff
@@ -0,0 +1,358 @@
+from snuba.clusters.cluster import ClickhouseCluster
+from snuba.datasets.slicing import is_storage_set_sliced
+from snuba.datasets.storages.factory import get_storage
+from snuba.datasets.storages.storage_key import StorageKey
+from snuba.pipeline.query_pipeline import QueryPipelineData, QueryPipelineStage
+from snuba.query import ProcessableQuery
+from snuba.query.composite import CompositeQuery
+from snuba.query.data_source.join import IndividualNode, JoinClause, JoinVisitor
+from snuba.query.data_source.simple import Table
+from snuba.query.data_source.visitor import DataSourceVisitor
+from snuba.query.query_settings import QuerySettings
+from snuba.querylog.query_metadata import (
+    QueryStatus,
+    SnubaQueryMetadata,
+    get_request_status,
+)
+from snuba.reader import Reader
+from snuba.settings import MAX_QUERY_SIZE_BYTES
+from snuba.utils.metrics.gauge import Gauge
+from snuba.utils.metrics.timer import Timer
+from snuba.utils.metrics.wrapper import MetricsWrapper
+from snuba.web import (
+    QueryException,
+    QueryExtraData,
+    QueryResult,
+    QueryTooLongException,
+    transform_column_names,
+)
+from snuba.web.db_query import db_query, update_query_metadata_and_stats
+
+metrics = MetricsWrapper(environment.metrics, "api")
+logger = logging.getLogger("snuba.pipeline.stages.query_execution")
+
+
+class StorageKeyJoinFinder(JoinVisitor[StorageKey, Table]):
+    """
+    Produces all the viable ClickhouseQueryPlans for each subquery
+    in the join.
+    """
+
+    def visit_individual_node(self, node: IndividualNode[Table]) -> StorageKey:
+        if isinstance(node.data_source, ProcessableQuery):
+            return node.data_source.get_from_clause().storage_key
+        else:
+            return node.data_source.storage_key
+
+    def visit_join_clause(self, node: JoinClause[Table]) -> StorageKey:
+        left_storage_key = node.left_node.accept(self)
+        right_storage_key = node.right_node.accept(self)
+        if is_storage_set_sliced(get_storage(left_storage_key).get_storage_set_key()):
+            return left_storage_key
+        return right_storage_key
+
+
+class StorageKeyFinder(DataSourceVisitor[StorageKey, Table]):
+    """
+    Given a query, finds the storage_set_key from which to get the cluster
+    In the case of a join, it will select an arbitrary storage_set_key that is in
+    the from_clause. Storages that are sliced are given higher priority
+
+    """
+
+    def _visit_simple_source(self, data_source: Table) -> StorageKey:
+        return data_source.storage_key
+
+    def _visit_join(self, data_source: JoinClause[Table]) -> StorageKey:
+        return data_source.accept(StorageKeyJoinFinder())
+
+    def _visit_simple_query(self, data_source: ProcessableQuery[Table]) -> StorageKey:
+        return data_source.get_from_clause().storage_key
+
+    def _visit_composite_query(self, data_source: CompositeQuery[Table]) -> StorageKey:
+        return self.visit(data_source.get_from_clause())
```

Comment:
This is the only new functionality! Before we depended on the StorageQueryPlanBuilder to select the cluster for us, which gives us a `Reader`. No more! Now because of this we can get the cluster straight from the physical query.,